File: src/p76/Solution.java
public class Solution {
    public String minWindow(String S, String T) {
        int ls = S.length(), lt = T.length();
        int cT[] = new int[256],fT[] = new int[256];
        for(int i = 0 ; i < lt; i++)
        	cT[T.charAt(i)]++;
        int hasFound = 0;
        int winStart = -1, winEnd = ls;
        for(int i = 0, start = 0 ; i < ls; i++) {
        	if(cT[S.charAt(i)] != 0) {
        		fT[S.charAt(i)]++;
        		if(fT[S.charAt(i)] <= cT[S.charAt(i)]) hasFound++;
        		if(hasFound == lt) {
        			while(cT[S.charAt(start)] == 0 || fT[S.charAt(start)] > cT[S.charAt(start)]) {
        				if(cT[S.charAt(start)] != 0)
        					fT[S.charAt(start)]--;
        				start++;
        			}
        			if(winEnd - winStart > i - start) {
        				winStart = start;
        				winEnd = i;
        			}
        			fT[S.charAt(start)]--;
        			start++;
        			hasFound--;
        		}
        	}
        }
        return winStart != -1 ? S.substring(winStart, winEnd + 1) : "";
    }
}

File: src/p76/solution.py
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        import collections
        ds = collections.defaultdict(int)
        dt = collections.defaultdict(int)
        has_found = 0
        for c in t:
            dt[c] += 1
        left = 0
        start, end = -1, len(s)
        for i, c in enumerate(s):
            if c in dt:
                ds[c] += 1
                if ds[c] <= dt[c]:
                    has_found += 1
                if has_found == len(t):
                    while s[left] not in dt or dt[s[left]] < ds[s[left]]:
                        if s[left] in dt:
                            ds[s[left]] -= 1
                        left += 1
                    if end - start > i - left:
                        start, end = left, i
                    ds[s[left]] -= 1
                    left += 1
                    has_found -= 1
        return "" if start == -1 else s[start:end + 1]
